### ISSUE TITLE
Hotkeys with modifiers currently don't have preference?

### ISSUE THREAD

**Issue by Tiax** [mezz/JustEnoughItems#517]
Could we please have hotkeys with modifiers (Ctrl, Shift, Alt, etc.) take preference over hotkeys without?

Currently you cannot map `[R]` to "Show Recipes" and `[Shift]+[R]` to "Show Uses", as the [InputHandler](https://github.com/mezz/JustEnoughItems/blob/1.10/src/main/java/mezz/jei/input/InputHandler.java#L232) does not take modifier keys into consideration!

**Comment by mezz** [comment]
Thanks for the report!

This is a current limitation of the Forge hotkey modifiers. `R` and `Shift+R` are considered conflicting hotkeys, because of the way modifiers are used in-game. If you hold shift to crouch, you can still press `E` to open your inventory. This means that a `Shift+E` hotkey is in conflict with `E`. 

Normally these conflicts are not a problem, so they are allowed. However in this case you are using conflicting hotkeys in a place where it is ambiguous, so JEI just picks one.

I think the solution here is I should improve the Forge hotkey system so that it works differently in GUIs than in-world. GUIs can always respect the modifiers, there's no worry about crouching and moving at the same time or anything like that.

**Comment by Tiax** [comment]
@mezz true- but I think the problem is entirely how JEI currently handles the input (code linked above). 

Say for example you have **no** bind for - let's say `(Alt) + R` - it's perfectly fine to have it handle `(Shift) + R` as a fallback for the regular `R` hotkey. As soon as you have `Alt+R` in the same context though, it should check that first, before falling back to the bind without modifiers.

In other words: it's probably smarter to check them in sequence, i.e. first check hotkeys with modifiers, then the ones without when no match found.

**Comment by mezz** [comment]
Fixed it with a PR to Forge, above. I wrote the Forge key modifier code btw.

This changes makes it so that `isActiveAndMatches` will be false for hotkey `R` when the `Shift` key is also held, and you're in a gui.
With the changes, JEI sees `showRecipe` as false and `showUses` as true when you press `Shift+R`.

**Comment by Tiax** [comment]
Thanks a lot for the rapid response+fix :)

Yet I'm not entirely sure this is desired behaviour though, i.e. check my example above.

Edit: it's also not how base-Minecraft handles it, e.g. Q and Shift+Q both perform the same thing (i.e. Shift+Q being not a valid bind).

**Comment by mezz** [comment]
I think it's preferable to do it this way because it makes the keybinds easier to use for modders, which means fewer bugs like this and more consistency. The case where you are holding `Alt` and also pressing an unrelated hotkey like `R` seems like a relatively rare case to me.

**Comment by Tiax** [comment]
Right - at the end of the day, it's all a matter of preference!

**Comment by mezz** [comment]
To answer your edit about dropping a stack:
The Minecraft keybinds are really strange and only partially comply with this system. There is no keybind in the controls for "Drop Selected Stack" like there should be, they implement it differently than a modder would. Thankfully this change does not break any of their hotkeys.

**Comment by Tiax** [comment]
> There is no keybind in the controls for "Drop Selected Stack" like there should be

Just to clarify: `Ctrl+Q` drops stack in the GUI, `Shift+Q` is the same as just pressing `Q` without any modifier, which is what I wanted to imply.

**Comment by mezz** [comment]
Ah okay. This still works actually, `Shift+Q` drops an item because vanilla doesn't use this system completely. I see what you're saying though.
